File: TaskManagementAPI/Entities/Models/Task.cs
using System.ComponentModel.DataAnnotations.Schema;

namespace Entities.Models
{
    public class Task
    {
        public Guid Id { get; set; } = Guid.NewGuid();
        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
        public long TaskSequence { get; set; }
        public long? ProjectId { get; set; }
        public string? CreatedById { get; set; }
        public string? AssignedToId { get; set; }
        public string Title { get; set; } = null!;
        public string? Description { get; set; }
        public TaskStatus Status { get; set; } = TaskStatus.ToDo;
        public TaskPriority Priority { get; set; } = TaskPriority.Medium;
        public DateTime? StartDate { get; set; }
        public DateTime? DueDate { get; set; }
        public DateTime? CompletedAt { get; set; }
        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
        public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;
        public decimal? EstimatedHours { get; set; }
        public decimal TotalHoursSpent { get; set; } = 0;
        public int ProgressPercentage { get; set; } = 0;
        public DateTime? DeletedAt { get; set; }
        [NotMapped]
        public bool IsDeleted => DeletedAt.HasValue;
        public Project? Project { get; set; }
        public Account? CreatedBy { get; set; }
        public Account? AssignedTo { get; set; }
        public ICollection<Comment> Comments { get; set; } = new List<Comment>();
        public ICollection<Label> Labels { get; set; } = new List<Label>();
        public ICollection<TaskAttachment> Attachments { get; set; } = new List<TaskAttachment>();
        public ICollection<TimeLog> TimeLogs { get; set; } = new List<TimeLog>();
        public ICollection<ActivityLog> ActivityLogs { get; set; } = new List<ActivityLog>();
    }

    public enum TaskStatus
    {
        ToDo = 0,
        InProgress = 1,
        InReview = 2,
        Done = 3,
        Blocked = 4,
    }

    public enum TaskPriority
    {
        Low = 0,
        Medium = 1,
        High = 2,
        Urgent = 3
    }
}

File: TaskManagementWeb/src/types/task.ts
import type Label from "./label";

export interface Task {
    id: string;
    createdByEmail: string;
    assignedToEmail: string;
    title: string;
    labelId?: string | null;
    status: TaskStatus;
    priority: TaskPriority;
    progressPercentage: number;
    commentCount: number;
    attachmentCount: number;
}

export interface TaskDetails {
    id: string;
    createdByEmail: string;
    createdByFirstName: string;
    createdByLastName: string;
    assignedToEmail: string;
    assignedToFirstName: string;
    assignedToLastName: string;
    labelName : string;
    labelColor : string;
    title: string;
    description: string | null;
    status: TaskStatus;
    priority: TaskPriority;
    startDate: string | null;
    dueDate: string | null;
    completedAt: string | null;
    createdAt: string;
    updatedAt: string;
    estimatedHours: number | null;
    totalHoursSpent: number;
    progressPercentage: number;
}

export interface TaskUpdateDTO {
    id: string;
    assignedToId?: string | null;
    labelId?: string | null;
    description?: string | null;
    title: string;
    status: TaskStatus;
    priority: TaskPriority;
    startDate?: string | null;
    dueDate?: string | null;
    completedAt?: string | null;
}

export const TaskStatus = {
    ToDo: "ToDo",
    InProgress: "InProgress",
    InReview: "InReview",
    Done: "Done",
    Blocked: "Blocked",
} as const;

export type TaskStatus = typeof TaskStatus[keyof typeof TaskStatus];

export const TaskPriority = {
    Low: "Low",
    Medium: "Medium",
    High: "High",
    Urgent: "Urgent",
} as const;

export type TaskPriority = typeof TaskPriority[keyof typeof TaskPriority];

export interface KanbanColumn {
  id: string;
  label: Label;
  tasks: Task[];
}

const priorityOrder: Record<TaskPriority, number> = {
    "Low": 0,
    "Medium": 1,
    "High": 2,
    "Urgent": 3
};

export const getPriorityOrder = (priority:  TaskPriority): number => {
    return priorityOrder[priority];
};

export const getPriorityLabel = (priority: TaskPriority): string => {
    const labels:  Record<TaskPriority, string> = {
        "Low":  "Düşük",
        "Medium":  "Orta",
        "High": "Yüksek",
        "Urgent": "Acil"
    };
    return labels[priority];
};

export const getStatusLabel = (status: TaskStatus): string => {
    const labels: Record<TaskStatus, string> = {
        "ToDo": "Yapılacak",
        "InProgress": "Devam Ediyor",
        "InReview": "İnceleme",
        "Done": "Tamamlandı",
        "Blocked": "Bloklandı"
    };
    return labels[status];
};